File: ivay-backend/src/main/java/com/ivay/service/CategoryService.java
package com.ivay.service;

import java.util.List;

import com.ivay.dtos.categorydto.CategoryRequestDto;
import com.ivay.dtos.categorydto.CategoryResponseDto;
import com.ivay.dtos.productdto.ProductResponseDto;

/**
 * Service interface for managing product categories.
 *
 * Provides methods to:
 * - retrieve all categories or by name
 * - retrieve, create, update, and delete a single category
 * - fetch products belonging to a specific category
 *
 * @since 1.0.0
 */
public interface CategoryService {

    /**
     * Retrieves all categories in the system.
     *
     * @return a list of {@link CategoryResponseDto} representing all categories
     */
    List<CategoryResponseDto> getAllCategories();

    /**
     * Finds categories whose name contains the given substring, case-insensitive.
     *
     * @param name substring to search for within category names
     * @return a list of matching {@link CategoryResponseDto}
     */
    List<CategoryResponseDto> findCategoriesByName(String name);

    /**
     * Retrieves a single category by its identifier.
     *
     * @param categoryId the identifier of the category to retrieve
     * @return the {@link CategoryResponseDto} for the given ID
     */
    CategoryResponseDto getCategoryById(Long categoryId);

    /**
     * Creates a new category.
     *
     * @param categoryRequestDto the data for the new category
     * @return the created {@link CategoryResponseDto}
     */
    CategoryResponseDto createCategory(CategoryRequestDto categoryRequestDto);

    /**
     * Updates an existing category.
     *
     * @param categoryId the identifier of the category to update
     * @param categoryRequestDto the new data for the category
     * @return the updated {@link CategoryResponseDto}
     */
    CategoryResponseDto updateCategory(Long categoryId, CategoryRequestDto categoryRequestDto);

    /**
     * Deletes a category by its identifier.
     *
     * @param categoryId the identifier of the category to delete
     */
    void deleteCategory(Long categoryId);

    /**
     * Retrieves all products associated with a specific category.
     *
     * @param categoryId the identifier of the category
     * @return a list of {@link ProductResponseDto} for products in that category
     */
    List<ProductResponseDto> getProductsByCategoryId(Long categoryId);
}

File: ivay-frontend/src/components/service/category.service.js
import axios from "axios";
import { API_BASE_URL } from "../constants/api";

const CATEGORY_API_URL = `${API_BASE_URL}/categories`;

/**
 * Fetches all categories.
 * @returns {Promise<Array>} A promise that resolves to an array of category response DTOs.
 */
const getAllCategories = async () => {
    const response = await axios.get(CATEGORY_API_URL);
    return response.data.data; // Assuming your ApiResponseDto wraps data
};

/**
 * Fetches categories whose names contain the given string.
 * @param {string} name - The name to filter by.
 * @returns {Promise<Array>} A promise that resolves to an array of category response DTOs.
 */
const findCategoriesByName = async (name) => {
    const response = await axios.get(`${CATEGORY_API_URL}/filter`, { params: { name } });
    return response.data.data;
};

/**
 * Fetches a category by its ID.
 * @param {number|string} categoryId - The ID of the category.
 * @returns {Promise<Object>} A promise that resolves to a category response DTO.
 */
const getCategoryById = async (categoryId) => {
    const response = await axios.get(`${CATEGORY_API_URL}/${categoryId}`);
    return response.data.data;
};

/**
 * Creates a new category.
 * @param {Object} categoryRequestDto - The category data to create.
 * @returns {Promise<Object>} A promise that resolves to the created category response DTO.
 */
const createCategory = async (categoryRequestDto) => {
    const response = await axios.post(CATEGORY_API_URL, categoryRequestDto);
    return response.data.data;
};

/**
 * Updates an existing category.
 * @param {number|string} categoryId - The ID of the category to update.
 * @param {Object} categoryRequestDto - The updated category data.
 * @returns {Promise<Object>} A promise that resolves to the updated category response DTO.
 */
const updateCategory = async (categoryId, categoryRequestDto) => {
    const response = await axios.put(`${CATEGORY_API_URL}/${categoryId}`, categoryRequestDto);
    return response.data.data;
};

/**
 * Deletes a category by its ID.
 * @param {number|string} categoryId - The ID of the category to delete.
 * @returns {Promise<void>} A promise that resolves when deletion is successful.
 */
const deleteCategory = async (categoryId) => {
    await axios.delete(`${CATEGORY_API_URL}/${categoryId}`);
    // No specific data returned on 204 No Content
};

/**
 * Fetches products associated with a specific category ID.
 * @param {number|string} categoryId - The ID of the category.
 * @returns {Promise<Array>} A promise that resolves to an array of product response DTOs.
 */
const getProductsByCategoryId = async (categoryId) => {
    const response = await axios.get(`${CATEGORY_API_URL}/${categoryId}/products`);
    return response.data.data;
};


const CategoryService = {
    getAllCategories,
    findCategoriesByName,
    getCategoryById,
    createCategory,
    updateCategory,
    deleteCategory,
    getProductsByCategoryId,
};

export default CategoryService;